docs/specification/README.md

```diff
@@ -191,18 +221,226 @@ The attachment index is an index to named attachments within the file. One recor
 | 4 + N | content_type | String | MIME type of the attachment. |
 | 8 | offset | uint64 | Byte offset to the attachment, from the start of the file. |
 
-#### Statistics (op=0x0A)
+### Statistics (op=0x08)
 
-The statistics record contains statistics about the recorded data. It is the last record in the file before the footer. The record must be preceded in the index data section by Channel Info records for any channels referenced in the `channel_message_counts` field. If this is undesirable but some statistics are still desired, the field may be set to a zero-length map. The statistics record is optional.
+A Statistics record contains summary information about the recorded data.
 
 | Bytes | Name | Type | Description |
 | --- | --- | --- | --- |
 | 8 | message_count | uint64 | Number of messages in the file across all topics. |
 | 4 | channel_count | uint32 | Number of channels in the file across all topics. |
 | 4 | attachment_count | uint32 | Number of attachments in the file. |
 | 4 | chunk_count | uint32 | Number of chunks in the file. |
-| N | channel_message_counts | KeyValues<uint16, uint64> | Mapping from channel ID to total message count for the channel. Duplicate keys are not allowed. |
+| 4 + N | channel_message_counts | Map<uint16, uint64> | Mapping from channel ID to total message count for the channel. An empty map indicates this statistic is not available. |
+
+When using a Statistics record with channel_message_counts, the Summary Data section MUST contain a copy of all Channel Info records. The Channel Info records MUST occur prior to the statistics record.
+
+> Why? The typical usecase for tools is to provide a listing of the types and quantities of messages stored in the file. Without an easy to access copy of the Channel Info records, tools would need to linearly scan the file for Channel Info records to display what types of messages exist in the file.
+
+### Metadata (op=0x09)
+
+A metadata record records arbitrary key/value user data.
+
+| Bytes | Name | Type | Description |
+| --- | --- | --- | --- |
+| 4 + N | metadata | Array<Tuple<string, string>> | Example keys: robot_id, git_sha, timezone, run_id. |
+
+### Summary Offset (op=0x0A)
+
+A Summary Offset record contains the location of a group of records with the same opcode within the summary section.
```

Comment:
Is it allowed to have more than one summary offset record with the same group opcode? Does the order matter?

Reply:
Order does not matter. If you split your groups of records then you would need more than one summary offset record. But doing this would degrade performance for little reason. We might want to add language preventing that. I think there is language about this in the Summary section.